apps/sms/js/thread_ui.js

```diff
@@ -688,38 +688,40 @@ var ThreadUI = {
     var findByAddress = Contacts.findByAddress.bind(Contacts);
     var number = activity.number;
     if (activity.contact && !number) {
-      findByAddress = function dummySource(contact, cb) {
-        cb(activity.contact);
-      };
+      findByAddress = () => Promise.resolve(activity.contact);
       number = activity.contact.number || activity.contact.tel[0].value;
     }
 
     // Add recipients and fill+focus the Compose area.
+    var recipientPromise;
     if (activity.contact && number) {
-      Utils.getContactDisplayInfo(
-        findByAddress, number, function onData(data) {
+      recipientPromise = Utils.getContactDisplayInfo(findByAddress, number)
+        .then(function onData(data) {
           data.source = 'contacts';
-          ThreadUI.recipients.add(data);
-          Compose.fromMessage(activity);
-        }
-      );
-    } else {
-      if (number) {
-        // If the activity delivered the number of an unknown recipient,
-        // create a recipient directly.
-        this.recipients.add({
-          number: number,
-          source: 'manual'
+          return data;
         });
-      }
-      Compose.fromMessage(activity);
-    }
-
-    if (number) {
-      Compose.focus();
+    } else if (number) {
+      // If the activity delivered the number of an unknown recipient,
+      // create a recipient directly.
+      recipientPromise = Promise.resolve({
+        number: number,
+        source: 'manual'
+      });
     } else {
-      this.recipients.focus();
+      recipientPromise = Promise.resolve();
     }
+
+    return recipientPromise.then((data) => {
+      data && this.recipients.add(data);
+
+      Compose.fromMessage(activity);
+
+      if (number) {
+        Compose.focus();
+      } else {
+        this.recipients.focus();
+      }
+    });
```

Comment:
I definitely OK with this! But honestly I don't see any value in ```Utils.getContactDisplayInfo```, it's used only here, internally it basically does nothing except that it calls public ```Utils.basicContact``` that's why I got rid of it in my patch (including useless tests, because in reality ```Utils.basicContact``` should be tested instead). + we use this approach in ThreadUI.handleDraft :) Do you think it's still useful?